### ISSUE TITLE
Error in Chrome (possibly others) from getDataFromHtml.js

### ISSUE THREAD

**Issue by mikeebee** [gmrchk/swupjs#16]
Hi there, 

This is a weird one and I'd love to try and figure out the root cause.

A production site has just become unusable for what appears to be only Chrome users on MacOS, which is what I saw the error on, that would have me think that it was a browser update but I have also a report that it affected Firefox on PC. I verified and diagnosed on `Chrome Version 73.0.3683.75 (Official Build) (64-bit)`

###  Error
`Uncaught TypeError: Cannot read property 'className' of null`

Which referenced this file https://github.com/gmrchk/swup/blob/master/src/modules/getDataFromHtml.js

From line 23

`pageClass: fakeDom.querySelector('#swupBody').className`

Which explicitly needs a Swup id of `swupBody`. I was using the default of `#swup`.

### Fix, sort of

I managed to fix it by changing my id to `swupBody` and specifying the `elements` option in `options`.

However, I can't use any other ID. If I don't use `#swupBody` is breaks.

There hasn't been a change to my sites codebase since last week so I'm a little baffled what has occurred.

I see some reference in [the docs](https://github.com/gmrchk/swup/blob/ba922a1a1530fb2b37259eacca8ceb80fd6938cb/readme.md#request-header) towards needing to alter the `getDataFromHtml.js` file but I'm not clear on how it references my use case.

### How I'm using Swup

Nothing fancy, using Greensock for some simple animations. If you need any more details please let me know.

### Update

Now I've made this change I'm seeing pages in Safari which attempt to load but just puts this in the address bar:

`typeerror:%20null%20is%20not%20an%20object%20(evaluating%20'document.body.querySelector('[data-swup="'%20+%20i%20+%20'"]').outerHTML%20=%20page.blocks[i]')`

But this may only be loosely related as there are JS errors on some of the pages from an external source, would that cause Swup to totally crash? I have disabled Swup on the production site for the time being.

I very much doubt this the plugins issue by the way! But I'd love some help finding out what is going on.

Many thanks!
Mike

**Comment by gmrchk** [comment]
Hi Mike, sorry for the late reply. 

Hope you solved your issue. `#swupBody` is definitely not something to dig into tho, as it's used by swup internally. This project is now deprecated in favor of [JS plugin](https://swup.js.org/plugins/js-plugin), so in case the problem still persists with JS plugin, let's continue on that repo.

**Comment by jp-prieto** [comment]
@gmrchk @mikeebee , all at the sudden I started experiencing this exact same issue on one of my sites. I've been using swup for well over a year.

After a lot of painful digging I've found that Swup.enable() will fail if any script injected in the head of the document references the string `<body`; 

the getDataFromHtml function uses the str.replace function to change the captured html's `<body` tag to a `<div id="swupBody"`. 

This is very fragile, if like in my case a third party app updated their script to now include a `<body` reference like the line `doc.write('<body onload="aFunction()"')` which shopify now uses to load their analytics engine then we are toast. 

getDataFromHtml is expecting only one `<body` on the html input and if the 1st occurrence is not the actual body then this error happens.

Here's the offending line in shopify's trekkie analytics script.

<img width="488" alt="Screen Shot 2019-09-05 at 12 43 27 AM" src="https://user-images.githubusercontent.com/40368988/64300813-43accf00-cf76-11e9-99bd-3d02f4b8c7c7.png">

A quick solution could be to maybe provide an extra option to enable() where we can specify the actual body string that should be parsed. It could default to `<body` but we could perhaps set it to say `<body data-swup-body` and add that data attribute on the body.

thanks for the hard work!

P.S - I'm using the latest swup version 2.0.5

**Comment by jp-prieto** [comment]
**OUTDATED** - check last post for cleaner solution.

I can confirm that this small changes  fixed my particular version of the className exception.

`getDataFromHtml.js`
<img width="534" alt="Screen Shot 2019-09-05 at 1 23 34 AM" src="https://user-images.githubusercontent.com/40368988/64302549-cedc9380-cf7b-11e9-9390-034657420290.png">

`getDataFromHtml call` 
<img width="684" alt="Screen Shot 2019-09-05 at 1 24 04 AM" src="https://user-images.githubusercontent.com/40368988/64302578-ea479e80-cf7b-11e9-818d-d754e9e27adc.png">

`Swup class default options`
<img width="309" alt="Screen Shot 2019-09-05 at 1 24 21 AM" src="https://user-images.githubusercontent.com/40368988/64302579-ea479e80-cf7b-11e9-9623-d4fabbb9ed93.png">

`Swup instance options`
<img width="281" alt="Screen Shot 2019-09-05 at 1 27 05 AM" src="https://user-images.githubusercontent.com/40368988/64302682-4d393580-cf7c-11e9-863b-fc926f412e50.png">

`html body`
<img width="218" alt="Screen Shot 2019-09-05 at 1 24 46 AM" src="https://user-images.githubusercontent.com/40368988/64302592-f92e5100-cf7b-11e9-95b9-d6a07d65ab84.png">

**Comment by mikeebee** [comment]
This sounds like what was happening to me, I was (unavoidably) using one or two external scripts that I had no control over and seemingly a little volatile, and Swup would just crash completely out of nowhere and make the site unusable.

I ended up having to remove it from the site 😔 I would _love_ to put it back in!

@gmrchk, does this error make sense from your point of view?

**Comment by jp-prieto** [comment]
Here's a codepen illustrating the issue:
https://codepen.io/createtribejp/pen/aboVgWo
Debug mode:
https://s.codepen.io/createtribejp/debug/aboVgWo/dGrXWKWQozaM

**Comment by jp-prieto** [comment]
@gmrchk, settling for this updated/cleaner solution.

This version safely executes the `id="swupBody"` replacement only within the body element, which should guarantee no head `<body` script references mess up the parsing. Tested it and so far so good, on both xhr and initial html.

`getDataFromHtml.js - v1.9.0 enhanced`
<img width="594" alt="Screen Shot 2019-09-06 at 12 36 29 AM" src="https://user-images.githubusercontent.com/40368988/64390862-67d3e300-d03e-11e9-9e57-7640e71867d8.png">

Hoe that helps.

JP

**Comment by mikeebee** [comment]
@jp-prieto just came to this issue to hopefully see about an update. I just re-read the full issue and @gmrchk said that this project is now deprecated and to continue over at the new repo. 

I'm not sure I will dare use this again until this issue has been addressed as it caused me huge issues on a site.

@jp-prieto I think it would be worth you opening a new issue/PR on that repo, you cool to do that? - https://github.com/swup/js-plugin

**Comment by marioloncarek** [comment]
Does anyone have an update on this? This bug crashed our shopify sites using swup.

**Comment by marioloncarek** [comment]
If anyone is using Shopify and having the same problems, here is how i managed to get it work: 

1) copied all Swup js files from node_modules to my JS folder (so i can change the core files)
2) initialised Swup from new folder 
3) opened getDataFromHtml.js 
4) added the code that @jp-prieto (thanks!) wrote (here is the non screenshot code so you can copy)

```
'use strict';

var _utils = require('./utils');

module.exports = function(html, request) {

    let tmpHTML = document.createElement('html');
    tmpHTML.innerHTML = html;

    let body = tmpHTML.querySelector('body').outerHTML;

    let content = body.replace('<body', '<div id="swupBody"')
        .replace('</body>', '</div>');

    let fakeDomHTML = tmpHTML.innerHTML.replace(body, content);
    let fakeDom = document.createElement('div');
    fakeDom.innerHTML = fakeDomHTML;
    let blocks = [];

    for (var i = 0; i < this.options.elements.length; i++) {
        if (fakeDom.querySelector(this.options.elements[i]) === null) {
            console.warn('Element ' + this.options.elements[i] + ' is not found in cached page.');
            return null;
        } else {
            (0, _utils.queryAll)(this.options.elements[i])
                .forEach((item, index) => {
                    (0, _utils.queryAll)(this.options.elements[i], fakeDom)[index].dataset.swup = blocks.length;
                    blocks.push((0, _utils.queryAll)(this.options.elements[i], fakeDom)[index].outerHTML);
                });
        }
    }

    var json = {
        title: fakeDom.querySelector('title').innerText,
        pageClass: fakeDom.querySelector('#swupBody').className,
        originalContent: html,
        blocks: blocks,
        responseURL: request != null ? request.responseURL : window.location.href,
    };
    return json;
};

```


That worked on our case, but it seems like the transitions are not perfect anymore, like they are laggy. But at least its working. Im planning to contact Shopify about that. Thanksa again @jp-prieto

**Comment by arnoldoosterom** [comment]
@marioloncarek I can confirm Shopify having this issue. Using 2.0.7. The workaround isn't working either but that's probably my fault. Sorry, I can't contribute any more than this.

**Comment by gmrchk** [comment]
Hi guys, sorry for the delay, but as stated, this is a deprecated repo, so it's little out of my sight. 

I admit the use of `replace` wasn't the cleanest idea, but I'm not sure it's actually the issue here. Anyway, who would think that a page would be loaded without a body tag? 🤷‍♂ 

Referencing [this issue](https://github.com/Shopify/slate/issues/1055), it seems that swup is basically running at the time when `body` is not present, which would also fit the description of the problem (`Uncaught TypeError: Cannot read property 'className' of null`). Initializing swup after the page has loaded and Shopify scripts executed would fix it in that case I guess, but I would need to check. Can someone confirm, or point me to someplace where the issue can be seen? Maybe some demo site? Seeing the actual Shopify analytics script would help too. 

Also, I would really love to hear the reasoning behind this line of code...
```javascript
doc.write('<body onload="document._boomrl();">')
```

@marioloncarek your workaround is probably laggy because you're creating a new `html` element on each transition as well as a `div` with lots of contents. Swup was only creating the div, so I guess the overhead wasn't that noticeable, but more importantly, Swup v2 [does a cleanup as well](https://github.com/swup/swup/blob/master/src/helpers/getDataFromHtml.js).

**Comment by arnoldoosterom** [comment]
Hi @gmrchk,

I've copied a theme and added the basic version of Swup. It produces the error for me using Chrome.
~~url: https://marijke-mul.myshopify.com~~
~~pass: swupjs~~
Removed

I couldn't find a unminified version of Shopify's Trekkie: https://cdn.shopify.com/s/javascripts/tricorder/trekkie.storefront.min.js?v=2017.09.05.1

It's my first time working with Swup so forgive rookie mistakes.

```
import 'components/jquery.module.js'; // https://stackoverflow.com/questions/35358625/jquery-is-not-defined-when-use-es6-import

import SmoothScroll from 'smoothscroll-for-websites';

import Swup from 'swup';
import SwupDebugPlugin from '@swup/debug-plugin';
import SwupBodyClassPlugin from '@swup/body-class-plugin';
import SwupJsPlugin from '@swup/js-plugin';
import SwupHeadPlugin from '@swup/head-plugin';
import SwupPreloadPlugin from '@swup/preload-plugin';

import animations from 'components/animation';
import sliders from 'components/sliders';
import overlays from 'components/overlay';
import utilities from 'components/utilities';

import account from 'components/account';
import filter from 'components/filter';


import shop from 'components/shop';


const swup = new Swup({
  bodySelector: 'data-swup-body',
  plugins: [
    new SwupDebugPlugin(),
    new SwupBodyClassPlugin(),
    new SwupHeadPlugin(),
    new SwupPreloadPlugin(),
    new SwupJsPlugin()
  ],
  cache: false,
  animateHistoryBrowsing: true
});

const loadPage = () => {

  animations.init()
  sliders.init()
  overlays.init()
  utilities.init()
  shop.init()
  account.init()
  filter.init()

}

document.addEventListener('DOMContentLoaded', loadPage)

// On page switch with Swup

swup.on('contentReplaced', function(){
  loadPage();
  window.scrollTo(0,0);
  return false;
});
```

**Comment by gmrchk** [comment]
Hi everyone, 

@arnoldoosterom thanks for the demo! I see what's the problem now. Quite a stupid mistake on my side... update swup to the current version (v2.0.8) and everything should be fixed. Please, let me know if you run into some related problems.

Also, I quickly went through the Shopify scripts, and if it was me who was a paying customer of Shopify, I would probably contact them about those scripts. Rewriting a global `document.write` probably works for them, but I still don't see the reason and mainly... who does that?! 🤷‍♂ 

Thanks for your input everyone!
